### PR TITLE
Add integration with MA.

### PR DESCRIPTION
**Pull request by jortel**
Initial integration with MA (xavier) validation service.
Adds:
- _Xavier_ client.
- A `revision` property to all resources.
- A `revisionAnalyzed` property which indicates the `revision` reflected in the analytics `concerns`.
- Triggered by _watch_ on vSphere Cluster and VM database entities.  Likely add a _watch_ on other kinds (Eg: Network) in the next iteration.

Need operator to set:
- XAVIER_URL
- XAVIER_USER
- XAVIER_PASSWORD

 in the virt-controller configMap.  Related PR (TBD).  Re: [issue/43](https://github.com/konveyor/virt-operator/issues/43).

Also, includes /vendor change for _common_ controller lib fix.

Requires: https://github.com/konveyor/controller/pull/34

### CODE REVIEW

File: pkg/controller/provider/container/ocp/collection.go [hunk 1]
```diff
@@ -53,7 +53,7 @@ func (r *StorageClass) Reconcile(ctx context.Context) (err error) {
 		m.With(&resource)
 		r.Reconciler.UpdateThreshold(m)
 		Log.Info("Create", libref.ToKind(m), m.String())
-		err = db.Insert(m)
+		err = tx.Insert(m)
```
Comment (jortel) [inline pkg/controller/provider/container/ocp/collection.go:56]: User transaction per _common_ controller change.

File: pkg/controller/provider/container/vsphere/reconciler.go [hunk 2]
```diff
@@ -375,30 +375,31 @@ next:
 			continue next
 		}
 		req.Version = updateSet.Version
-		transaction, err = r.db.Begin()
```
Comment (jortel) [inline pkg/controller/provider/container/vsphere/reconciler.go:378]: just to be consistent.

File: pkg/controller/provider/model/vsphere/model.go [hunk 3]
```diff
@@ -23,8 +28,8 @@ type Base struct {
 	Name string `sql:""`
 	// Parent
 	Parent string `sql:"index(a)"`
-	// Annotations
-	Annotations string `sql:""`
```
Comment (jortel) [inline pkg/controller/provider/model/vsphere/model.go:27]: Annotations did not end up being generally useful.
Replaced with `concerns` on VM.

File: pkg/controller/provider/model/vsphere/model.go [hunk 4]
```diff
@@ -49,6 +58,26 @@ func (m *Base) Equals(other libmodel.Model) bool {
...
+}
+
+//
+// Updated.
+// Increment revision. Should ONLY be called by
+// the reconciler.
+func (m *Base) Updated() {
+	m.Revision++
+}
+
```
Comment (jortel) [inline pkg/controller/provider/model/vsphere/model.go:80]: Methods for managing the revision.

File: pkg/controller/provider/xavier/base/client.go [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+		"2",
+		"value",
+		"org.drools.core.runtime.rule.impl.FlatQueryResults",
+		"idResultMaps",
+		"element",
+		"0",
+		"element",
+		"0",
+		"value",
+		"org.jboss.xavier.analytics.pojo.output.workload.inventory.WorkloadInventoryReportModel")
```
Comment (jortel) [inline pkg/controller/provider/xavier/base/client.go:124]: The path into the returned REST resource is absurd.

File: pkg/controller/provider/xavier/base/client.go [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+	if len(concerns) == 0 {
+		recommended := false
+		if list, cast := providers.(List); cast {
+			for _, provider := range list {
+				if provider == "OCP" {
+					recommended = true
+					break
+				}
+			}
+		}
```
Comment (jortel) [inline pkg/controller/provider/xavier/base/client.go:154]: Rules are triggered by _known_ field values.  So if, for example, an unknown guest OS is passed, some of the rules don't fire resulting in `OCP` not being in the recommendTargetsIMS but flagsIMS = nil.  This added as a catch-all.

File: pkg/controller/provider/xavier/base/client.go [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+			return
+		}
+	}
+
+	return
+}
+
+//
+// Get object at path.
+func (r Client) find(in interface{}, path ...string) (out interface{}, err error) {
```
Comment (jortel) [inline pkg/controller/provider/xavier/base/client.go:288]: Needed to safely traverse the absurd REST payloads.

File: pkg/controller/provider/xavier/vsphere/adapter.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+import (
+	liberr "github.com/konveyor/controller/pkg/error"
+	libmodel "github.com/konveyor/controller/pkg/inventory/model"
+	model "github.com/konveyor/virt-controller/pkg/controller/provider/model/vsphere"
+	"github.com/konveyor/virt-controller/pkg/controller/provider/xavier/base"
+)
+
+//
+// Model adapter.
```
Comment (jortel) [inline pkg/controller/provider/xavier/vsphere/adapter.go:11]: Model adapter is the same pattern used in the _container_ package.

File: pkg/controller/provider/xavier/base/client.go [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+	header.Add(
+		"Authorization",
+		"Basic "+encoded)
+
+	return
+}
+
+//
+// Post validation request.
+func (r *Client) post(in, out interface{}) (status int, err error) {
```
Comment (mansam) [inline pkg/controller/provider/xavier/base/client.go:247]: I only see this method called in one place; is it necessary to use an interface type for these parameters? It looks like at the very least `out` will always be a `base.Object{}`.

File: pkg/controller/provider/xavier/base/client.go (hunk 7 above)
Comment (mansam) [reply pkg/controller/provider/xavier/base/client.go:288]: Likewise, `in` and `out` here both look like they are only ever a `base.Object`.

File: pkg/controller/provider/xavier/vsphere/adapter.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+type Adapter struct {
+	// DB
+	libmodel.DB
+}
+
+//
+// Update the vSphere xavier body.
+func (r *Adapter) UpdateBody(object base.Object, m libmodel.Model) (err error) {
```
Comment (mansam) [inline pkg/controller/provider/xavier/vsphere/adapter.go:19]: This returns an error, but its only caller ignores the error. Is that an oversight, or deliberate?

File: pkg/controller/provider/xavier/base/client.go [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+	}
+	client := http.Client{}
+	response, err := client.Do(request)
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	status = response.StatusCode
+	content := []byte{}
+	if status == http.StatusOK {
```
Comment (mansam) [inline pkg/controller/provider/xavier/base/client.go:269]: This appears to be ignoring statuses other than 200. Is there going to be other error handling added later?

**Review by mansam — COMMENTED** [review]
Looks good, just a small handful of questions.

File: pkg/controller/provider/xavier/base/client.go (hunk 9 above)
Comment (jortel) [reply pkg/controller/provider/xavier/base/client.go:247]: Yes, the post() can be narrowed to an `*Object` for both in and out.

File: pkg/controller/provider/xavier/base/client.go (hunk 7 above)
Comment (jortel) [reply pkg/controller/provider/xavier/base/client.go:288]: Not really.  Depending on the path, the `out` could be a `List`, string or anything.   The `out` is used as `in` in other places so even for `in` to be typed as `*Object` would make things difficult.

File: pkg/controller/provider/xavier/vsphere/adapter.go (hunk 10 above)
Comment (jortel) [reply pkg/controller/provider/xavier/vsphere/adapter.go:19]: Oversight.  Fixed.

File: pkg/controller/provider/xavier/base/client.go (hunk 11 above)
Comment (jortel) [reply pkg/controller/provider/xavier/base/client.go:269]: No.  Anything besides a 200 is an error.

**Review by mansam — APPROVED** [review]

**Pull request merged**